tests/matmul/generate_e2e_matmul_tests.py

```diff
@@ -0,0 +1,624 @@
+    name: str,
+    matrix_shape: list,
+    element_type: MatrixElemTypeId,
+):
+    global pseudorandom_generator_seed
+    pseudorandom_generator_seed = pseudorandom_generator_seed + 1
+    return (
+        f"  %{name}_dim0 = arith.constant {matrix_shape[0]} : i64\n"
+        f"  %{name}_dim1 = arith.constant {matrix_shape[1]} : i64\n"
+        f"  %{name}_element_type = hal.element_type<{element_type.value}> : i32\n"
+        f"  %{name}_seed = arith.constant {pseudorandom_generator_seed} : i32\n"
+        f"  %{name} = call @matmul_test.generate_random_matrix(%device, %{name}_dim0, %{name}_dim1, %{name}_element_type, %{name}_seed) : (!hal.device, i64, i64, i32, i32) -> !hal.buffer_view\n"
+    )
+
+
+call_id = 0
+
+
+# Generates the output trace for a testcase i.e. a single test function call,
+# as a dictionary to be passed to yaml.dump.
+def generate_call(
+    function: MLIRFunction,
+    lhs_rhs_type: MatrixElemTypeId,
+    acc_type: MatrixElemTypeId,
+    shape: TestShape,
+):
+    global call_id
+    func_name = f"{function.name}_{shape.m}_{shape.k}_{shape.n}"
+    if shape.accumulate:
+        func_name = f"{func_name}_acc"
+    func_name = f"{func_name}_{call_id}"
+    call_id = call_id + 1
+
+    description = f"Matmul shape (MxKxN): {shape.m}x{shape.k}x{shape.n}"
+    op = (
+        f"func.func @{func_name}() attributes {{\n"
+        f'  iree.reflection = {{description = "{description}"}}\n'
+        "} {\n"
+        "  %device_index = arith.constant 0 : index\n"
+        "  %device = hal.devices.get %device_index : !hal.device\n"
+    )
+
+    op = op + generate_random_matrix("lhs", [shape.m, shape.k], lhs_rhs_type)
+    op = op + generate_random_matrix("rhs", [shape.k, shape.n], lhs_rhs_type)
+    if shape.accumulate:
+        op = op + generate_random_matrix("acc", [shape.m, shape.n], acc_type)
+        # TODO(#16168): there's a bug with in-place input->output aliasing and
+        # we work around it here by passing in a unique copy.
+        global pseudorandom_generator_seed
+        pseudorandom_generator_seed = pseudorandom_generator_seed - 1
+        op = op + generate_random_matrix("acc_copy", [shape.m, shape.n], acc_type)
+        op = op + (
+            f"  %result = call @module.{function.name}(%lhs, %rhs, %acc_copy) : (!hal.buffer_view, !hal.buffer_view, !hal.buffer_view) -> !hal.buffer_view\n"
+        )
+    else:
+        op = op + (
+            f"  %acc = util.null : !hal.buffer_view\n"
+            f"  %result = call @module.{function.name}(%lhs, %rhs) : (!hal.buffer_view, !hal.buffer_view) -> !hal.buffer_view\n"
+        )
+
+    op = op + (
+        f"  %m = arith.constant {shape.m} : i64\n"
+        f"  %k = arith.constant {shape.k} : i64\n"
+        f"  %n = arith.constant {shape.n} : i64\n"
+        f"  call @matmul_test.check_matmul_results(%device, %m, %k, %n, %lhs, %rhs, %acc, %result) : (!hal.device, i64, i64, i64, !hal.buffer_view, !hal.buffer_view, !hal.buffer_view, !hal.buffer_view) -> ()\n"
```

Comment:
If only we had python bindings that enabled us to build such ir using an in-memory representations of ops and types and the module................ lol

Reply:
We do and use it for some things. But strings are easy :)

Reply:
I guess it says something about the bindings that strings are _still_ easier.

Reply:
Meh, strings are always easier for this level of thing... Virtually no cognitive load and everyone can hack on them.

Remind me some day to dig up some of the early web assembly predecessor asm "string work" :)